### DISCUSSION TITLE
Proposal: Language feature "return if myCondition"

### DISCUSSION THREAD

**Discussion by dibley1973** [dotnet/csharplang#1493] [Language Ideas]
I am a fan of one line guard clauses like.

```
Ensure.IsNotNull(myArgument, nameof(myArgument))
````
Which would throw an exception if the argument value is null.

So in the same vein, my currently **most desired new C# feature** would be a "**return if**" kind of statement.
```
var myCondition = DoSomeChecks(...);
return if myCondition;
```

Which would return from a method if the condition is true. It would simply be _syntatic-sugar_ for

```
var myCondition = DoSomeChecks(...);
if (myCondition)
{
    return;
}
```

Is this a language feature which could be considered for inclusion in a future version of C#?

**Comment by alrz** [comment]
What would be the advantage of this?
```cs
return if (myCondition);
if (myCondition) return;

təˈmeɪtoʊ;
təˈmɑːtəʊ;
```

**Comment by ufcpp** [comment]
gaurd statement? https://github.com/dotnet/csharplang/issues/138

**Comment by dibley1973** [comment]
@alrz - because some companies (like where I work) enforce curly braces for statements for the condition, even if its a one liner. If added to the language as per my post then people working for companies which require these code rules can have more succinct code.

Having to have the braces means your guard clauses can take up half your screen!

**Comment by dibley1973** [comment]
@ufcpp - nice to see others have already proposed it.

**Comment by alrz** [comment]
>because some companies (like where I work) enforce curly braces for statements for the condition, 

Riot.

Seriously though, you want this feature to get around your company's coding convention? that *really* doesn't seem to be a good reason to add a language feature IMO. I'd expect that such strict guidelines to forbid using `return if` constructs to avoid inconsistencies. So you'd be back to where you started. I'm pretty sure your company requires blocks for a reason, no?

**Comment by jnm2** [comment]
See #325, #453, and #724 for existing discussion.

**Comment by yaakov-h** [comment]
Well then,

```csharp
return if (myCondition);
if (myCondition) { return; }
```

As much as I like the idea of a `guard` statement, is an entirely new language feature really worth reordering a construct to save a handful of characters?

**Comment by magol** [comment]
But if the function must return a value, how do you do that with this proposed solution?

**Comment by asyncawaitmvvm** [comment]
guard, defer, return if... these all go into file 13.  My friend in high school joked that we should have an "unless" statement. Throw that in there too.

**Comment by yaakov-h** [comment]
That's #560. 😄

**Comment by Mteheran** [comment]
Take a Look to the following code:

```
public decimal? CalculateNewPlannedAmountForLastPaidLine(EscalationDataAmount escalationDataAmount, decimal? rootLineTotalPlanned, DateTime? plannedDate)
        {
            if (!rootLineTotalPlanned.HasValue)
            {
                return null;
            }

            decimal newPlanned = CalculateNewPlannedAmount(escalationDataAmount, plannedDate);

            return newPlanned - rootLineTotalPlanned.Value;
        }
```

Right now we cannot mix the first condition (if (!rootLineTotalPlanned.HasValue)) 
 on the last return. I think this could be useful in this case.

**Comment by CyrusNajmabadi** [comment]
You can already write this as:

```c#
public decimal? CalculateNewPlannedAmountForLastPaidLine(EscalationDataAmount escalationDataAmount, decimal? rootLineTotalPlanned, DateTime? plannedDate)
    => rootLineTotalPlanned.HasValue
        ? CalculateNewPlannedAmount(escalationDataAmount, plannedDate) - rootLineTotalPlanned.Value
        : default(decimal?);
```

I think there are already enough language features to help get brevity here.  I don't think a new form of 'return' is really needed.

**Comment by asyncawaitmvvm** [comment]
> That's #560. 😄

I don't know who you are or how you found that but you made my day.
